File: android/app/src/main/java/com/example/shilv/service/PlaceNameService.kt
package com.example.shilv.service

import android.content.Context
import android.location.Geocoder
import com.example.shilv.data.GeoPoint
import java.util.Locale

/**
 * 地点名称解析服务：对应 iOS CLGeocoder。
 *  - 缓存避免重复请求
 *  - 仅按需解析有限坐标
 */
class PlaceNameService(private val context: Context) {

    data class PlaceName(
        val shortName: String,
        val city: String?,
        val country: String?,
    )

    private val cache = mutableMapOf<String, PlaceName?>()
    private var lastRequestAt: Long = 0L

    suspend fun resolve(point: GeoPoint): PlaceName? {
        val key = cacheKey(point)
        if (cache.containsKey(key)) return cache[key]
        respectRateLimit()
        return try {
            val geocoder = Geocoder(context, Locale.CHINA)
            val marks = geocoder.getFromLocation(point.latitude, point.longitude, 1)
            lastRequestAt = System.currentTimeMillis()
            if (marks.isNullOrEmpty()) { cache[key] = null; return null }
            val mark = marks.first()
            val city = mark.locality ?: mark.subAdminArea ?: mark.adminArea
            val short = mark.featureName ?: mark.subLocality ?: city ?: mark.countryName ?: "旅途中的一站"
            val value = PlaceName(shortName = short, city = city, country = mark.countryName)
            cache[key] = value
            value
        } catch (t: Throwable) {
            lastRequestAt = System.currentTimeMillis()
            cache[key] = null
            null
        }
    }

    suspend fun locate(query: String): GeoPoint? {
        if (query.isBlank()) return null
        respectRateLimit()
        return try {
            val geocoder = Geocoder(context, Locale.CHINA)
            val marks = geocoder.getFromLocationName(query, 1)
            lastRequestAt = System.currentTimeMillis()
            if (marks.isNullOrEmpty()) return null
            val mark = marks.first()
            GeoPoint(mark.latitude, mark.longitude)
        } catch (t: Throwable) {
            lastRequestAt = System.currentTimeMillis()
            null
        }
    }

    private fun cacheKey(point: GeoPoint) =
        "${(point.latitude * 100).toLong() / 100.0},${(point.longitude * 100).toLong() / 100.0}"

    private fun respectRateLimit() {
        val elapsed = System.currentTimeMillis() - lastRequestAt
        if (elapsed < 250) {
            try { Thread.sleep(250 - elapsed) } catch (_: InterruptedException) { /* ignore */ }
        }
    }
}

File: ios/ShiLv/Services/PlaceNameService.swift
import Foundation
@preconcurrency import CoreLocation

actor PlaceNameService {
    private var cache: [String: PlaceName] = [:]
    private var lastRequestAt: Date?

    struct PlaceName: Sendable {
        let shortName: String
        let city: String?
        let country: String?
    }

    func resolve(_ point: GeoPoint) async -> PlaceName? {
        let key = cacheKey(point)
        if let cached = cache[key] { return cached }
        await respectRateLimit()
        let geocoder = CLGeocoder()
        do {
            let marks = try await geocoder.reverseGeocodeLocation(CLLocation(latitude: point.latitude, longitude: point.longitude), preferredLocale: Locale(identifier: "zh_CN"))
            guard let mark = marks.first else { return nil }
            let city = mark.locality ?? mark.subAdministrativeArea ?? mark.administrativeArea
            let short = mark.name ?? mark.subLocality ?? city ?? mark.country ?? "旅途中的一站"
            let value = PlaceName(shortName: short, city: city, country: mark.country)
            cache[key] = value
            lastRequestAt = Date()
            return value
        } catch {
            lastRequestAt = Date()
            return nil
        }
    }

    func locate(_ query: String) async -> GeoPoint? {
        guard !query.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return nil }
        await respectRateLimit()
        do {
            let marks = try await CLGeocoder().geocodeAddressString(query, in: nil, preferredLocale: Locale(identifier: "zh_CN"))
            lastRequestAt = Date()
            guard let coordinate = marks.first?.location?.coordinate else { return nil }
            return GeoPoint(latitude: coordinate.latitude, longitude: coordinate.longitude)
        } catch { lastRequestAt = Date(); return nil }
    }

    private func cacheKey(_ point: GeoPoint) -> String {
        "\((point.latitude * 100).rounded() / 100),\((point.longitude * 100).rounded() / 100)"
    }

    private func respectRateLimit() async {
        guard let lastRequestAt else { return }
        let remaining = 250_000_000 - Int64(Date().timeIntervalSince(lastRequestAt) * 1_000_000_000)
        if remaining > 0 { try? await Task.sleep(nanoseconds: UInt64(remaining)) }
    }
}
